Google.Api.Gax.Grpc/Rest/PartialDecodingStreamReader.cs

```diff
@@ -43,117 +101,147 @@ public PartialDecodingStreamReader(Task<TextReader> textReaderTask, Func<string,
+            }
 
-            if (resultTask == cancellationTask)
+            // Throwing an exception takes priority over marking the sequence as completed.
+            // But if there were valid responses *before* we detected an error, we'll return those first.
+            if (_errorMessage is string)
             {
-                // If the cancellationTask "wins" `Task.WhenAny` by being cancelled, the following await will throw TaskCancelledException.
-                await cancellationTask.ConfigureAwait(false);
+                MaybeDispose();
+                // TODO: Should this actually be an RpcException?
+                throw new InvalidOperationException(_errorMessage);
             }
 
-            var readLen = await taskRead.ConfigureAwait(false);
-            if (readLen == 0)
+            if (_endOfData)
             {
-                var errorText = "Closing `]` bracket not received after iterating through the stream. " +
-                                "This means that streaming ended without all objects transmitted. " +
-                                "It is likely a result of server or network error.";
-                throw new InvalidOperationException(errorText);
+                MaybeDispose();
+                return false;
             }
 
-            var readChars = buffer.Take(readLen);
-            foreach (var c in readChars)
+            // We don't currently know what to do, so read more data and see whether that
+            // satisfies one of the above conditions. (We may need to loop multiple times.)
+            await ReadAndProcessData(cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    /// <summary>
+    /// Called when we don't have a queued response and haven't reached a terminal state (error or completed).
+    /// </summary>
+    /// <returns></returns>
+    private async Task ReadAndProcessData(CancellationToken cancellationToken)
+    {
+        try
+        {
+            _textReader ??= await AwaitWithCancellation(_textReaderTask, cancellationToken, defaultValue: null).ConfigureAwait(false);
+            if (_textReader is null)
             {
-                // Closing bracket for the top-level array
-                if (_currentBuffer.Length == 0 && c == ']')
-                {
-                    // TODO[virost, jskeet, 11/2022] Fix with tokenizer:
-                    // it's possible to receive more data after the closing `]`
-                    // We currently just close the reader, which means we won't detect a response which
-                    // is bad due to trailing non-whitespace after the ]... it's unclear whether it's
-                    // worth continuing to read just to catch that though.
-                    _arrayClosed = true;
-                    _textReader.Close();
-                    // We can't just return false here, as we might have queued responses.
-                    break;
-                }
+                // Cancellation
```

Comment:
Or the task result is a null TextReader, in which case, you'll be here forever, or not, you'll be here until actual cancellation where the problem really was that the TextReader was null.

This will possibly never happen depending on how we construct this.

Reply:
Right - changing to throw in AwaitWithCancellation will somewhat fix this; we'll get an NRE if we ever create a task that returns null, but that can't actually happen.